table/src/main/java/com/bazaarvoice/emodb/table/db/astyanax/CQLStashTableDAO.java

```diff
@@ -0,0 +1,307 @@
+                                .and(QueryBuilder.eq("is_start_token", true))
+                                .limit(1)
+                                .setConsistencyLevel(ConsistencyLevel.LOCAL_QUORUM));
+
+        Row row = resultSet.one();
+
+        if (row == null) {
+            return Iterators.emptyIterator();
+        }
+
+        TableJson tableJson = toTableJson(row.getString(0));
+        return Iterators.singletonIterator(new ProtoStashTokenRange(fromInclusive, toExclusive, tableJson));
+    }
+
+    private Iterator<ProtoStashTokenRange> getTokenRangesBetweenInterShard(String stashId, String placement,
+                                                                           ByteBuffer fromInclusive, ByteBuffer toExclusive) {
+        // Since the range crosses the boundary between at least two tables it is guaranteed that if any table has a shard
+        // whose token range intersects the query range then by querying for all start and end tokens within that range
+        // either the start token, end token, or both will be returned in the results.
+
+        ResultSet resultSet = _placementCache.get(_systemTablePlacement)
+                .getKeyspace()
+                .getCqlSession()
+                .execute(
+                        QueryBuilder.select("range_token", "is_start_token", "table_json")
+                                .from(STASH_TOKEN_RANGE_TABLE)
+                                .where(QueryBuilder.eq("stash_id", stashId))
+                                .and(QueryBuilder.eq("data_center", _dataCenters.getSelf().getName()))
+                                .and(QueryBuilder.eq("placement", placement))
+                                .and(QueryBuilder.gte("range_token", fromInclusive))
+                                .and(QueryBuilder.lt("range_token", toExclusive))
+                                .setConsistencyLevel(ConsistencyLevel.LOCAL_QUORUM)
+                                .setFetchSize(100));
+
+        final Iterator<Row> resultSetIterator = resultSet.iterator();
+
+        return new AbstractIterator<ProtoStashTokenRange>() {
+            TableJson currentTable;
+            ByteBuffer currentTableStartToken;
+
+            @Override
+            protected ProtoStashTokenRange computeNext() {
+                ProtoStashTokenRange range = null;
+
+                while (range == null) {
+                    if (resultSetIterator.hasNext()) {
+                        Row row = resultSetIterator.next();
+                        TableJson table = getTableJson(row);
+                        if (isStartToken(row)) {
+                            if (currentTable == null) {
+                                currentTable = table;
+                                currentTableStartToken = getToken(row);
+                            } else {
+                                throw new IllegalStateException("Overlapping table range start rows found");
```

Comment:
Yup we get this if we encounter two table range start rows. But how are we avoiding?

Reply:
This is a condition which should never be met.  It's not that we're avoiding it; the data written by `addTokenRangesForTable()` should never make it in the first place.  If this happens it means that the read storage itself is returning overlapped rows for tables, which would cause much bigger problems than this.  That's why it's throwing `IllegalArgumentException`, if this ever happens it is a stop-the-world kind of error.